### DISCUSSION TITLE
Init cluster error on master role

### DISCUSSION THREAD

**Discussion by nawarajshahi** [techno-tim/k3s-ansible#184] [Questions] [answered]
hey @timothystewart6, 
I am currently trying to deploy this to my ubuntu 22.04 vms and I keep getting the undefined variable errors. 

## Expected Behavior
master node role  should copy the templates to destination as noted on the roles 

## Current Behavior
This should copy the templates to destination as noted on the roles but it fails to template the files under templates folder for master role. 

##error 
```
TASK [k3s/master : Init cluster inside the transient k3s-init service] **************************************************************************************************************************
fatal: [192.168.33.119]: FAILED! => 
  msg: |-
    The task includes an option with an undefined variable. The error was: {% if groups['master'] | length > 1 %}
      {% if ansible_hostname == hostvars[groups['master'][0]]['ansible_hostname'] %}
        --cluster-init
      {% else %}
        --server https://{{ hostvars[groups['master'][0]].k3s_node_ip }}:6443
      {% endif %}
      --token {{ k3s_token }}
    {% endif %} {{ extra_server_args | default('') }}: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'k3s_node_ip'
  
    The error appears to be in '/home/nabu/repos/projects/k3s-ansible/roles/k3s/master/tasks/main.yml': line 62, column 3, but may
    be elsewhere in the file depending on the exact syntax problem.
  
    The offending line appears to be:
  
  
    - name: Init cluster inside the transient k3s-init service
      ^ here
fatal: [192.168.33.120]: FAILED! => 
  msg: |-
    The task includes an option with an undefined variable. The error was: {% if groups['master'] | length > 1 %}
      {% if ansible_hostname == hostvars[groups['master'][0]]['ansible_hostname'] %}
        --cluster-init
      {% else %}
        --server https://{{ hostvars[groups['master'][0]].k3s_node_ip }}:6443
      {% endif %}
      --token {{ k3s_token }}
    {% endif %} {{ extra_server_args | default('') }}: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'k3s_node_ip'
  
    The error appears to be in '/home/nabu/repos/projects/k3s-ansible/roles/k3s/master/tasks/main.yml': line 62, column 3, but may
    be elsewhere in the file depending on the exact syntax problem.
  
    The offending line appears to be:
  
  
    - name: Init cluster inside the transient k3s-init service
      ^ here
```

## Steps to Reproduce

changed all the required lines under hosts.ini and all.yaml 

## Context (variables)
ubuntu 22.04 running on proxmox vm

##Question
Do you have any suggestions for resolving this issue?

**Comment by timothystewart6** [comment, marked as answer]
be sure to check https://github.com/techno-tim/k3s-ansible/discussions/20  This works on 22.04, we test it in CI with 22.04

**Reply by nawarajshahi** [reply]
thanks for the suggestion @timothystewart6 and also, a huge thanks for putting all this together. I had used this a couple of months go and was successfully able to deploy HA k3s cluster. I decided to give it a go today and had some hiccups. 

Thanks again!

**Comment by pramitbiswas** [comment]
I get an error in the exact location.

hostvars[groups['master'][0]] is fine, but for other master IP (i.e., else block: `--server https://{{ hostvars[groups['master'][0]].k3s_node_ip | split(",") | first | ansible.utils.ipwrap }}:6443`) getting the error. `k3s_node_ip` (i.e., 192.168.30.39) probably not in the `hostvars[groups['master'][0]]`. 

```
The task includes an option with an undefined variable. The error was: {% if groups['master'] | length > 1 %}\n  {% if ansible_hostname == hostvars[groups['master'][0]]['ansible_hostname'] %}\n    --cluster-init\n  {% else %}\n    --server https://{{ hostvars[groups['master'][0]].k3s_node_ip | split(\",\") | first | ansible.utils.ipwrap }}:6443\n  {% endif %}\n  --token {{ k3s_token }}\n{% endif %} {{ extra_server_args | default('') }}: Unable to look up a name or access an attribute in template string ({% if groups['master'] | length > 1 %}\n  {% if ansible_hostname == hostvars[groups['master'][0]]['ansible_hostname'] %}\n    --cluster-init\n  {% else %}\n    --server https://{{ hostvars[groups['master'][0]].k3s_node_ip | split(\",\") | first | ansible.utils.ipwrap }}:6443\n  {% endif %}\n  --token {{ k3s_token }}\n{% endif %} {{ extra_server_args | default('') }}).\nMake sure your variable name does not contain invalid characters like '-': descriptor 'split' for 'str' objects doesn't apply to a 'AnsibleUndefined' object
```

ansible [core 2.12.10]
python version = 3.9.5
nodes dist: Raspbian

**Comment by pramitbiswas** [comment]
@timothystewart6, can you provide an example/expected output of `--server https://{{ hostvars[groups['master'][0]].k3s_node_ip | split(",") | first | ansible.utils.ipwrap }}:6443`.

**Comment by actdiux** [comment]
I ended up here with the same error. I got past that error by changing 
```
# interface which will be used for flannel
flannel_iface: "enp2s0"
#flannel_iface: "eth0"
```
after I noticed this part of the error 
`--node-ip={{ k3s_node_ip }}: {{ ansible_facts[flannel_iface][\"ipv4\"][\"address\"] }}: 'dict object' has no attribute 'eth0'`

hope it helps the next person

**Reply by pramitbiswas** [reply]
Thanks, but in my case, all interfaces are `eth0`. Checked with `"{{ ansible_default_ipv4.interface }}"` as well as manually from `ifconfig`. Infact, I do not get error for creating `k3s_node_ip` var. Rather, `hostvars[groups['master'][0]]` `dict` has no attribute `k3s_node_ip` (i.e., the IP denoted by it).

**Comment by pramitbiswas** [comment]
`"{{ hostvars[groups['master'][0]].ansible_facts[flannel_iface]['ipv4']['address'] | split(',') | first | ansible.utils.ipwrap }}"` this is fine but not `{{ hostvars[groups['master'][0]].k3s_node_ip | split(",") | first | ansible.utils.ipwrap }}`.
